etest_test/profile_test.py

```diff
@@ -0,0 +1,153 @@
+"""Tests for etest.profile."""
+
+import logging
+from typing import Any, Dict
+
+import pytest
+from hypothesis import assume, given
+from hypothesis import strategies as st
+from pytest_mock import MockerFixture
+
+import etest.profile as sut
+
+logger = logging.getLogger(__name__)
+
+
+_valid_architectures = ["amd64", "x86", "armv5", "armv6", "armv7", "arm64", "ppc64"]
+_valid_systemd_architectures = ["amd64", "x86", "arm64", "ppc64"]
+_libc_value = ["glibc", "musl", "uclibc"]
+
+
+@st.composite
+def non_base_profiles(draw: Any) -> Dict[str, Any]:
+    """Strategy for generating non-base profiles."""
+    # Draw values
+    profile = {}
+    profile["libc"] = draw(st.sampled_from(_libc_value))
+    profile["hardened"] = draw(st.booleans())
+    profile["multilib"] = draw(st.booleans())
+    profile["systemd"] = draw(st.booleans())
+
+    # Ensure it isn't a base profile
+    assume(profile != {"libc": "glibc", "hardened": False, "multilib": True, "systemd": False})
+
+    return profile
+
+
+@pytest.mark.parametrize("arch", _valid_architectures)
+def test_base_arch(arch: str) -> None:
+    """Ensure the base profiles work."""
+    result = sut.Profile(False, arch, "glibc", False, True, False)
+
+    # Check profile strings
+    assert result.profile == arch
+
+
+@pytest.mark.parametrize("libc", ["musl", "uclibc"])
+@pytest.mark.parametrize("hardened", [True, False])
+def test_amd64_libc(libc: str, hardened: bool) -> None:
+    """Ensure alternative libcs work on amd64."""
+    arch = "amd64"
+
+    result = sut.Profile(True, arch, libc, hardened, True, False)
+
+    # Check the profile string
+    result_suffix = "hardened" if hardened else "vanilla"
+    assert result.profile == f"{arch}-{libc}-{result_suffix}"
+
+
+@pytest.mark.parametrize("arch", ["armv5", "armv6", "armv7"])
+@given(profile=non_base_profiles())
+def test_invalid_arm_profiles(arch: str, profile: Dict[str, Any]) -> None:
+    """Ensure non-base ARM profiles raise exceptions."""
+    with pytest.raises(sut.InvalidProfileError):
+        sut.Profile(True, arch, profile["libc"], profile["hardened"], profile["multilib"], profile["systemd"])
+
+
+@given(profile=non_base_profiles())
+def test_invalid_arm64_profiles(profile: Dict[str, Any]) -> None:
+    """Ensure invalid ARM64 profiles raise exceptions."""
+    arch = "arm64"
+
+    # Dont test arm64-systemd, as it is a valid profile
+    assume(profile != {"libc": "glibc", "hardened": False, "multilib": True, "systemd": True})
```

Comment:
Using assume should be avoided if at all possible as it's immensely slower and leads to bad samples taking up computation time.  This is one I'd start with parametrised and move on to property once a property is identified.

Reply:
I got the impression from the hypothesis docs that assume was the right thing to do for these kinds of tests. Using parametrize would lead to code complexity (and Im also not sure how to exclude this particular case from that implementation)

Reply:
You would simply create a new strategy that can't generate this particular case.  Also a tuple of booleans multiplied by a libc is a pretty straightforward strategy (especially to add exceptions to).  Take a peek at the Tuple generator and how you can feed that to a choice generator to include or exclude a choice.  If we're trying to prove negatives it gets tricker but should still have its own strategy just with the opposite logic from the correct profile (probably with arbitrary libc strings as well, but that seems like overkill).

Due to the low cardinality on these values I wouldn't use hypothesis.  Because it has low cardinality I would head for smallcheck (if it existed in python) or parametrise.  That's why I suggested dropping hypothesis and starting with parameterise.  We can fully declare the representative invalid space quite easily.